util/tracecpu/profile.go

```diff
@@ -0,0 +1,401 @@
+		s.plans[k] = v + (v/planTotal)*remain
+	}
+}
+
+func (sp *sqlStatsProfiler) handleExportProfileTask(p *profile.Profile) {
+	sp.mu.Lock()
+	defer sp.mu.Unlock()
+	if sp.mu.ept == nil {
+		return
+	}
+	sp.mu.ept.mergeProfile(p)
+}
+
+func (sp *sqlStatsProfiler) hasExportProfileTask() bool {
+	sp.mu.Lock()
+	has := sp.mu.ept != nil
+	sp.mu.Unlock()
+	return has
+}
+
+// IsEnabled return true if it is(should be) enabled. It exports for tests.
+func (sp *sqlStatsProfiler) IsEnabled() bool {
+	return config.GetGlobalConfig().TopSQL.Enable || sp.hasExportProfileTask()
+}
+
+// StartCPUProfile same like pprof.StartCPUProfile.
+// Because the GlobalSQLStatsProfiler keep calling pprof.StartCPUProfile to fetch SQL cpu stats, other place (such pprof profile HTTP API handler) call pprof.StartCPUProfile will be failed,
+// other place should call tracecpu.StartCPUProfile instead of pprof.StartCPUProfile.
+func StartCPUProfile(w io.Writer) error {
+	if GlobalSQLStatsProfiler.IsEnabled() {
+		return GlobalSQLStatsProfiler.startExportCPUProfile(w)
+	}
+	return pprof.StartCPUProfile(w)
+}
+
+// StopCPUProfile same like pprof.StopCPUProfile.
+// other place should call tracecpu.StopCPUProfile instead of pprof.StopCPUProfile.
+func StopCPUProfile() error {
+	if GlobalSQLStatsProfiler.IsEnabled() {
+		return GlobalSQLStatsProfiler.stopExportCPUProfile()
+	}
+	pprof.StopCPUProfile()
+	return nil
+}
+
+// SetGoroutineLabelsWithSQL sets the SQL digest label into the goroutine.
+func SetGoroutineLabelsWithSQL(ctx context.Context, normalizedSQL, sqlDigest string) context.Context {
+	if len(normalizedSQL) == 0 || len(sqlDigest) == 0 {
+		return ctx
+	}
+	if variable.EnablePProfSQLCPU.Load() {
+		ctx = pprof.WithLabels(context.Background(), pprof.Labels(labelSQLDigest, sqlDigest, labelSQL, util.QueryStrForLog(normalizedSQL)))
+	} else {
+		ctx = pprof.WithLabels(context.Background(), pprof.Labels(labelSQLDigest, sqlDigest))
+	}
+	pprof.SetGoroutineLabels(ctx)
+	GlobalSQLStatsProfiler.RegisterSQL(sqlDigest, normalizedSQL)
+	return ctx
+}
+
+// SetGoroutineLabelsWithSQLAndPlan sets the SQL and plan digest label into the goroutine.
+func SetGoroutineLabelsWithSQLAndPlan(ctx context.Context, sqlDigest, planDigest, normalizedPlan string) context.Context {
+	ctx = pprof.WithLabels(ctx, pprof.Labels(labelSQLDigest, sqlDigest, labelPlanDigest, planDigest))
+	pprof.SetGoroutineLabels(ctx)
+	GlobalSQLStatsProfiler.RegisterPlan(planDigest, normalizedPlan)
+	return ctx
+}
+
+func (sp *sqlStatsProfiler) startExportCPUProfile(w io.Writer) error {
+	sp.mu.Lock()
+	defer sp.mu.Unlock()
+	if sp.mu.ept != nil {
+		return errors.New("cpu profiling already in use")
+	}
+	sp.mu.ept = &exportProfileTask{w: w}
+	return nil
+}
+
+func (sp *sqlStatsProfiler) stopExportCPUProfile() error {
+	sp.mu.Lock()
+	ept := sp.mu.ept
+	sp.mu.ept = nil
+	sp.mu.Unlock()
+	if ept.err != nil {
+		return ept.err
+	}
+	if w := ept.w; w != nil && ept.cpuProfile != nil {
+		sp.removeLabel(ept.cpuProfile)
```

Comment:
Why the SQL label is removed? what will happen without this line?

Reply:
Already add a comment for removeLabel functions.